File: Demo/Objective_C_Demo/ViewController/ChatViewController.m
//
//  ChatViewController.m
//  Demo
//
//  Created by IEMacBook01 on 21/05/16.
//  Copyright © 2016 Iftekhar. All rights reserved.
//

#import "ChatViewController.h"
#import "ChatTableViewCell.h"

@interface ChatViewController ()<UITableViewDataSource,UITableViewDelegate,UITextFieldDelegate>

@property (strong, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation ChatViewController
{
    NSMutableArray<NSString*> *texts;
    
    IBOutlet UIButton *buttonSend;
    IBOutlet UITextField *inputTextField;
}

-(void)viewDidLoad
{
    [super viewDidLoad];
    
    texts = [[NSMutableArray alloc] initWithObjects:@"This is demo text chat. Enter your message and hit `Send` to add more chat.", nil];
    
    inputTextField.inputAccessoryView = [[UIView alloc] init];
}

-(void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textFieldDidChange:) name:UITextFieldTextDidChangeNotification object:inputTextField];
}

-(void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    [[NSNotificationCenter defaultCenter] removeObserver:self name:UITextFieldTextDidChangeNotification object:inputTextField];
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return texts.count;
}

-(CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 100;
}

-(CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return UITableViewAutomaticDimension;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    ChatTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ChatTableViewCell" forIndexPath:indexPath];
    cell.chatLabel.text = texts[indexPath.row];
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

- (IBAction)sendAction:(UIButton *)sender
{
    NSString *text = [inputTextField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    
    if (text.length != 0)
    {
        NSIndexPath *indexPath = [NSIndexPath indexPathForRow:[_tableView numberOfRowsInSection:0] inSection:0];

        [texts addObject:text];
        inputTextField.text = @"";
        buttonSend.enabled = NO;
        
        [self.tableView insertRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        [self.tableView scrollToRowAtIndexPath:indexPath atScrollPosition:UITableViewScrollPositionNone animated:YES];
    }
}

-(void)textFieldDidBeginEditing:(UITextField *)textField
{

}

-(void)textFieldDidChange:(NSNotification*)notification
{
    NSString *text = [inputTextField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    buttonSend.enabled = text.length != 0;
}

@end

File: Demo/Swift_Demo/ViewController/ChatViewController.swift
//
//  ChatViewController.swift
//  Demo
//
//  Created by IEMacBook01 on 23/05/16.
//  Copyright © 2016 Iftekhar. All rights reserved.
//


class ChatViewController: UIViewController, UITableViewDataSource,UITableViewDelegate,UITextFieldDelegate {

    @IBOutlet var tableView : UITableView!

    @IBOutlet var buttonSend : UIButton!
    @IBOutlet var inputTextField : UITextField!

    var texts = ["This is demo text chat. Enter your message and hit `Send` to add more chat."]
    
    override func viewDidLoad() {
        super.viewDidLoad()
        inputTextField.inputAccessoryView = UIView()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.textFieldDidChange(_:)),    name: NSNotification.Name.UITextFieldTextDidChange, object: inputTextField)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        NotificationCenter.default.removeObserver(self, name: NSNotification.Name.UITextFieldTextDidChange, object: inputTextField)
    }
    
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return texts.count
    }
    
    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        return 100
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return UITableViewAutomaticDimension
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "ChatTableViewCell", for: indexPath) as! ChatTableViewCell
        cell.chatLabel.text = texts[(indexPath as NSIndexPath).row]
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
    }
    
    @IBAction func sendAction(_ sender : UIButton) {
        if inputTextField.text?.characters.count != 0 {

            let indexPath = IndexPath(row: tableView.numberOfRows(inSection: 0), section: 0)
            
            texts.append(inputTextField.text!)
            inputTextField.text = ""
            buttonSend.isEnabled = false
            
            tableView.insertRows(at: [indexPath], with:UITableViewRowAnimation.automatic)
            tableView.scrollToRow(at: indexPath, at:UITableViewScrollPosition.none, animated:true)

        }
    }
    
    func textFieldDidChange(_ notification: Notification) {
        buttonSend.isEnabled = inputTextField.text?.characters.count != 0
    }
    
    func textFieldDidBeginEditing(_ textField: UITextField) {

    }
}
